Add flocker volume plugin

agonzalezro opened the pull request:
Flocker [1] is an open-source container data volume manager for
Dockerized applications.

This PR adds a volume plugin for Flocker.
The plugin interfaces the Flocker Control Service REST API [2] to
request volume creation and attachment.

Each kubelet host should run Flocker agents (Container Agent and Dataset
Agent).

The kubelet will also require environment variables that contain the
host and port of the Flocker Control Service. (see Flocker architecture
[3] for more).
- `FLOCKER_CONTROL_SERVICE_HOST`
- `FLOCKER_CONTROL_SERVICE_PORT`

The contribution introduces a new 'flocker' volume type to the API with
fields:
- `datasetName`: which indicates the name of the dataset in Flocker
  added to metadata;
- `size`: a human-readable number that indicates the maximum size of the
  requested dataset.

Full documentation (including getting started guides) and e2e tests to
follow.

As the plugin is code complete, we're hoping to make version 1.1.

[1] https://clusterhq.com/flocker/introduction/
[2] https://docs.clusterhq.com/en/1.3.1/reference/api.html
[3] https://docs.clusterhq.com/en/1.3.1/concepts/architecture.html

timothysc commented:
cc @rootfs @markturansky  - just FYI | review.

markturansky commented on pkg/volume/flocker/plugin.go:
```diff
@@ -0,0 +1,138 @@
+}
+
+func (p flockerPlugin) CanSupport(spec *volume.Spec) bool {
+	return (spec.PersistentVolume != nil && spec.PersistentVolume.Spec.Flocker != nil) ||
```
You can just `return spec.Volume != nil && spec.Volume.Flocker != nil`

This plugin is not a Persistent Volume, so you can ignore that part of the CanSupport check.

markturansky replied:
nm :) I overlooked Flocker in the PVVolumeSource struct.  Your check above is correct.

markturansky commented on pkg/volume/flocker/plugin.go:
```diff
@@ -0,0 +1,138 @@
+	if spec.Volume != nil && spec.Volume.Flocker != nil {
+		return spec.Volume.Flocker, readOnly
+	}
+	return spec.PersistentVolume.Spec.Flocker, readOnly
```
Not a PV.  This last line will panic.

markturansky replied:
Same as above.  I missed flocker in PV struct.  Looks good here.

thockin commented:
You guys are cutting this close...

thockin commented on pkg/api/v1/types.go:
```diff
@@ -306,6 +309,8 @@ type PersistentVolumeSource struct {
 	CephFS *CephFSVolumeSource `json:"cephfs,omitempty"`
 	// FC represents a Fibre Channel resource that is attached to a kubelet's host machine and then exposed to the pod.
 	FC *FCVolumeSource `json:"fc,omitempty"`
+	// Flocker represents a Flocker volume attached to a kubelet's host machine and exposed to the pod for its usage
```
Please also document that this depends on the flocker control plane being running.

thockin commented on pkg/api/validation/validation.go:
```diff
@@ -522,6 +527,22 @@ func validateGlusterfs(glusterfs *api.GlusterfsVolumeSource) errs.ValidationErro
+		allErrs = append(allErrs, errs.NewFieldInvalid("datasetName", flocker.DatasetName, "must not contain '/'"))
+	}
+	if flocker.Size != "" {
+		if _, err := units.FromHumanSize(flocker.Size); err != nil {
```
We should stay consistent with our Quantity type, here.

agonzalezro replied:
Sorry @thockin I don't know what you mean with your comment.

thockin replied:
we have a type we use in our API for human-readable quantities - `resource.Quantity` - if you make the API field be a quanitity then you can validate it without this 'units' library, and it supports common syntax with our other resource variants.

agonzalezro replied:
Oh cool, I am going to take a look to that one. Sorry, I didn't know it exists and I checked the Godeps to see what you had there.

thockin replied:
If this is to go in for v1.1 it needs to be approved and clear of
encumberances by EOD Friday.  Otherwise it's going to miss the boat.  Note
that this doesn't mean every nit is fixed, but that major structural issues
are resolved and small details will be fixed soon after in a followup.

On Wed, Sep 23, 2015 at 8:05 AM, Alexandre González <
notifications@github.com> wrote:

> In pkg/api/validation/validation.go
> <https://github.com/kubernetes/kubernetes/pull/14328#discussion_r40214231>
> :
>
> > @@ -522,6 +527,22 @@ func validateGlusterfs(glusterfs *api.GlusterfsVolumeSource) errs.ValidationErro
> >  	return allErrs
> >  }
> >
> > +func validateFlocker(flocker *api.FlockerVolumeSource) errs.ValidationErrorList {
> > +	allErrs := errs.ValidationErrorList{}
> > +	if flocker.DatasetName == "" {
> > +		allErrs = append(allErrs, errs.NewFieldRequired("datasetName"))
> > +	}
> > +	if strings.Contains(flocker.DatasetName, "/") {
> > +		allErrs = append(allErrs, errs.NewFieldInvalid("datasetName", flocker.DatasetName, "must not contain '/'"))
> > +	}
> > +	if flocker.Size != "" {
> > +		if _, err := units.FromHumanSize(flocker.Size); err != nil {
>
> Oh cool, I am going to take a look to that one. Sorry, I didn't know it
> exists and I checked the Godeps to see what you had there.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/14328/files#r40214231>.
>

thockin commented:
I took a quick look, nothing major jumped out.  I'll leave it to Mark and Saad to pick the details.

What this is missing most is a good doc explaining how to use it, including the bits about env variables in the first comment.

rootfs commented on pkg/volume/flocker/util.go:
```diff
@@ -0,0 +1,30 @@
+// otherwise it return the default value provided.
+//
+// Note: it should be GetEnv but I tried to keep the os.Getenv standard.
+func GetenvOrFallback(key, defaultValue string) string {
```
getenvOrFallback?

rootfs commented on pkg/volume/flocker/client.go:
```diff
@@ -0,0 +1,391 @@
+ * 3) If not, create a new one (we will need the Primary UUID)
+ * 4) Wait until the dataset is ready or the timeout was reached
+ */
+func (c flockerClient) CreateVolume(dir string) (datasetID string, err error) {
```
createVolume?

rootfs commented on pkg/volume/flocker/client.go:
```diff
@@ -0,0 +1,391 @@
+	if host == "" {
+		return nil, errFlockerControlServiceHost
+	}
+	portEnv := os.Getenv("FLOCKER_CONTROL_SERVICE_PORT")
```
no default?

agonzalezro replied:
Yeah, I think that you are right, I will default here as well.

rootfs commented on pkg/volume/flocker/client.go:
```diff
@@ -0,0 +1,391 @@
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
```
double close?

agonzalezro replied:
Why do you think so? `.get` calls `.request` and no one of those closes the body.

rootfs commented on pkg/volume/flocker/client.go:
```diff
@@ -0,0 +1,391 @@
+	}
+	datasetID = payload.DatasetID
+
+	// 4) Wait until the dataset is ready for usage, this can take a long a time
```
a long time?

rootfs commented on pkg/volume/flocker/client.go:
```diff
@@ -0,0 +1,391 @@
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
```
it appears caller of `getDatasetState` already remembers to close the connection.

rootfs commented on pkg/volume/flocker/plugin.go:
```diff
@@ -0,0 +1,138 @@
+		return err
+	}
+
+	datasetID, err := c.CreateVolume(dir)
```
flocker doesn't need to mount the volume? how does containers access to volume? if docker mounts the volume, is there any docker release dependency?

agonzalezro replied:
That's something the Flocker agent does. We know the path where Flocker decided to mount it with the `GetPath` function in the line 109

markturansky commented on cmd/kubelet/app/plugins.go:
```diff
@@ -54,21 +55,26 @@ func ProbeVolumePlugins() []volume.VolumePlugin {
+		flocker.ProbeVolumePlugins(),
+	} {
+		allPlugins = append(allPlugins, plugins...)
+	}
```
Why this change to an array and loop?  I had to do a doubletake to see if anything actually changed ..

agonzalezro replied:
I know that this could be a problem since I am changing lines not related but I just thought that it was going to be more readable this way. I can change it back if there is a problem.

markturansky replied:
A doubletake or re-parsing, to me, is less readable because I have to think about it.  

Plugins initialized elsewhere (there are several other places) all follow the original.

rootfs commented:
@agonzalezro Is flocker volume dependent on any docker release? Does it support rkt? Can you provide a working Pod and running environment in a doc?

rootfs commented on pkg/volume/flocker/plugin.go:
```diff
@@ -0,0 +1,138 @@
+	return &builder, nil
+}
+
+func (p *flockerPlugin) NewCleaner(datasetName string, podUID types.UID) (volume.Cleaner, error) {
```
you don't want to delete the volume?

agonzalezro replied:
The idea of Flocker is that you can move it between pods. I don't see at this stage any need for removing volume. Flocker agent will mount/dismount it when needed.

markturansky replied:
But when the volume is moved between pods and between hosts, don't you want to clean up the orphaned directory on the old host?

mattbates replied:
Not in this case, no. If and when the same dataset is requested elsewhere (by datasetName in the volume spec), the Flocker agent will umount/detach and reattach/mount the volume. If a volume is not required, at this stage it is a manual operation to destroy it. Hope that helps explain.

pmorie replied:
This makes me wonder if there is any utility in using Flocker as a PV, since the semantics seem intrinsic to how the Flocker agent handles volumes.

pmorie commented on pkg/volume/flocker/client.go:
```diff
@@ -0,0 +1,391 @@
+ * - FLOCKER_CONTROL_SERVICE_CLIENT_CERT_FILE
+ */
+func newFlockerClient(pod *api.Pod, useTLS bool) (*flockerClient, error) {
+	host := os.Getenv("FLOCKER_CONTROL_SERVICE_HOST")
```
I'd rather plumb these through from the kubelet than check for environment variables in a volume plugin long-term.

markturansky replied:
If this is configured once for the plugin and its value is correct across the cluster, then using VolumeConfig would allow CLI configuration of all these variables.

agonzalezro replied:
Yes, VolumeConfig looks a good option to consider. Env variables will work for now so perhaps this is one for a next iteration?

pmorie replied:
We need to discuss how this kind of configuration is going to be plumbed through to volume plugins.  See #14288

saad-ali replied:
@agonzalezro
See https://github.com/kubernetes/kubernetes/pull/9870 for example on how to pass through environment variables via VolumeConfig.

I'm ok with the environment variables as long as they are well documented.

mattbates replied:
We took a good look at VolumeConfig and agree it's the preferred way to go ultimately, but [OtherAttributes](https://github.com/kubernetes/kubernetes/blob/568c0331b02e9b655f28c6c60516d44a1389c6b2/pkg/volume/plugins.go#L219) doesn't appear to be implemented and this seemed to be the right place for these plugin-specific settings.. So our best option for now, that is implemented, tested and now documented (see latest commits), is to use env vars and consider VolumeConfig in a next iteration. Hope you agree?

saad-ali replied:
@mattbates Yep, we're ok with using environment variables for now. Thanks for adding the documentation. Please open up an issue to make sure that the work to switch to VolumeConfig is tracked.

markturansky replied:
OtherAttributes is just a map of strings.  You would add your vars to an
instance of VolumeConfig.OtherAttributes["yourvar"], you pass VolumeConfig
to your plugin, and your plugin understands what the keys and expected
values are.

It should work and be that easy.  If it's not, please let me know and I can
fix whatever you may find.

Saad previously said he was OK with documented ENV vars.  It would not take
much to port them (now or in the future, I have no preference) to
VolumeConfig.  I think it should be a smooth transition and anyone using
those vars shouldn't see any difference.

On Fri, Sep 25, 2015 at 1:24 PM, Matthew Bates <notifications@github.com>
wrote:

> In pkg/volume/flocker/client.go
> <https://github.com/kubernetes/kubernetes/pull/14328#discussion_r40453742>
> :
>
> > + * newFlockerClient creates a wrapper over http.Client to communicate with the
> > + * flocker control service. The location of this service is defined by the
> > + * following environment variables:
> > + *
> > + * - FLOCKER_CONTROL_SERVICE_HOST
> > + * - FLOCKER_CONTROL_SERVICE_PORT
> > + *
> > + * If you are using the TLS client you can also specify where the cert files
> > + * reside:
> > + *
> > + * - FLOCKER_CONTROL_SERVICE_CA_FILE
> > + * - FLOCKER_CONTROL_SERVICE_CLIENT_KEY_FILE
> > + * - FLOCKER_CONTROL_SERVICE_CLIENT_CERT_FILE
> > + */
> > +func newFlockerClient(pod *api.Pod, useTLS bool) (*flockerClient, error) {
> > +	host := os.Getenv("FLOCKER_CONTROL_SERVICE_HOST")
>
> We took a good look at VolumeConfig and agree it's the preferred way to go
> ultimately, but OtherAttributes
> <https://github.com/kubernetes/kubernetes/blob/568c0331b02e9b655f28c6c60516d44a1389c6b2/pkg/volume/plugins.go#L219>
> doesn't appear to be implemented and this seemed to be the right place for
> these plugin-specific settings.. So our best option for now, that is
> implemented, tested and now documented (see latest commits), is to use env
> vars and consider VolumeConfig in a next iteration. Hope you agree?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/14328/files#r40453742>.
>

agonzalezro commented:
@rootfs Flocker is not docker/rkt dependent. We are working in documentation for it. cc: @mattbates

lavalamp commented on pkg/api/v1/types.go:
```diff
@@ -589,6 +594,15 @@ type CephFSVolumeSource struct {
+	DatasetName string `json:"datasetName"`
+
+	// Optional: the size of the volume in human readable format.
+	Size string `json:"size"`
```
Please use resource.Quantity as this type. (Find other usages elsewhere in this file.)

Meta: it seems odd to specify a size here. Other volumes (including persistent volumes) don't specify a size. Why is this volume different?

markturansky replied:
You're right that the volume itself does not require this information.  

It's irrelevant to a Volume and size is already an attribute of PV.Spec.

saad-ali replied:
+1

lavalamp commented:
@dchen1107 Can you find a code reviewer for this? I think it's mostly node stuff.

dchen1107 commented:
cc/ @kubernetes/goog-cluster

bgrant0607 commented:
If you haven't already, please also take a look at:
https://github.com/kubernetes/kubernetes/blob/master/docs/devel/faster_reviews.md
https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api_changes.md
https://github.com/kubernetes/kubernetes/blob/master/docs/devel/coding-conventions.md

agonzalezro commented:
I’ve addressed all the comments I could, now we are investigating how to proceed with

1) using VolumeConfig vs env vars
2) size parameter
3) volume cleaning

saad-ali commented on pkg/volume/flocker/client.go:
```diff
@@ -0,0 +1,382 @@
+/*
```
Is there anything kubernetes specific in this file? If not, I feel like this should be a separate project (a flocker go client) that anyone can reference as a dependency.

pmorie replied:
+1

agonzalezro replied:
It was before, it isn't now, so move it out is going to be a good decision. Doing it just now.

saad-ali commented on pkg/volume/flocker/client.go:
```diff
@@ -0,0 +1,382 @@
+type flockerClient struct {
+	*http.Client
+
+	pod *api.Pod
```
Where is this (`pod *api.Pod`) actually used?

agonzalezro replied:
Good catch

saad-ali commented on pkg/volume/flocker/client.go:
```diff
@@ -0,0 +1,382 @@
+	return newFlockerClient(pod, true)
+}
+
+func newFlockerClientWithSize(pod *api.Pod, maximumSize int64) (*flockerClient, error) {
```
Could you please add GoDoc comments. I'd like to quickly be able to understand what the difference between `...WithSize` and `...WithoutSize` is, for example.

mattbates commented:
CLA has been signed but googlebot seems to have not picked it up.. we've re-pushed and  I've used my email address as signed in the CLA.

thockin commented:
I show @agonzalezro as CLA but not @mattbates yes

mattbates commented:
I've signed with my corp email address, now added to my GH account. What should I do to give this a kick? As I said, we've re-pushed already. Thanks.

thockin commented:
I don't see a corp CL for jetstack - when did you sign?

mattbates commented:
Signed 3 days ago. Just clicked the link in the email to sign and it says it's signed already.
Could it perhaps be an email alias issue? Have updated my commits with the author email address the same I signed the CLA, if it helps.

agonzalezro committed:
Add flocker volume plugin

Flocker [1] is an open-source container data volume manager for
Dockerized applications.

This PR adds a volume plugin for Flocker.
The plugin interfaces the Flocker Control Service REST API [2] to
attachment attach the volume to the pod.

Each kubelet host should run Flocker agents (Container Agent and Dataset
Agent).

The kubelet will also require environment variables that contain the
host and port of the Flocker Control Service. (see Flocker architecture
[3] for more).

- `FLOCKER_CONTROL_SERVICE_HOST`
- `FLOCKER_CONTROL_SERVICE_PORT`

The contribution introduces a new 'flocker' volume type to the API with
fields:

- `datasetName`: which indicates the name of the dataset in Flocker
  added to metadata;
- `size`: a human-readable number that indicates the maximum size of the
  requested dataset.

Full documentation can be found docs/user-guide/volumes.md and examples
can be found at the examples/ folder

[1] https://clusterhq.com/flocker/introduction/
[2] https://docs.clusterhq.com/en/1.3.1/reference/api.html
[3] https://docs.clusterhq.com/en/1.3.1/concepts/architecture.html

saad-ali commented on examples/flocker/README.md:
```diff
@@ -0,0 +1,128 @@
+
+#### Flocker Control Service
+
+The Flocker Control Service should be installed manually on a host outside (but accessible from) the Kubernetes cluster. In the future, this may be deployed in pod(s) and exposed as a Kubernetes service.
```
> Flocker Control Service should be installed manually on a host outside (but accessible from) the Kubernetes cluster

What exactly does that mean, and why does it have to be outside the Kubernetes cluster?

mattbates replied:
For now, it's recommended to be deployed as a standard daemon on a host (VM/bare metal - which could also be the K8s master for that matter). Just wanted to stress here that it's not a Kubernetes-managed pod/service. Happy to reword as necessary..

saad-ali replied:
Could be on a random node machine as well right?
It would be great if you could reword that. Thanks!

saad-ali commented on examples/flocker/README.md:
```diff
@@ -0,0 +1,128 @@
+
+- **datasetName** is the unique name for the Flocker dataset.
+
+If the dataset does not already exist in Flocker, it is automatically provisioned, attached and mounted by the plugin. Otherwise, with an existing dataset, it is reattached by Flocker to the node that the pod is scheduled.
```
Flocker should *not* auto-provision inside the `Builder` (i.e. as a result of `SetUp()`/`SetUpAt(...)` methods). The `Builder` should only "attach"/"mount" the volume (I understand the name is a bit misleading).

@markturansky is in the process of adding dynamic provisioning support. See https://github.com/kubernetes/kubernetes/pull/13650 and https://github.com/kubernetes/kubernetes/pull/14537. In short, the `Creater` interface should be the hook that results in volume provisioning.

thockin replied:
Yeah, auto-provisioning probably should not happen here.  With acknowledgement that flocker is a somewhat different model than we've seen in the past, I think the simpler thing to do right now is to error out if the volume does not exist, and then we can tie into the nascent provisioning infrastructure.

What think?

fabioy commented:
Looks like this PR is also getting hit by #14677 (Travis CI failure).

agonzalezro commented:
@fabioy I actually thought that it was my error and I was trying to fix it. Any workaround for now? Or should we wait for #14677 to be fixed?

fabioy commented:
No workarounds for now... Some unexplained behavior of the tool seems to be causing this.

agonzalezro commented:
Ok @fabioy, thanks for your support. I will be waiting for some new regarding the CLA & the `Godeps` issue then.

thockin commented:
I see a CLA for Jetstack now, but CLA bot is not clearing you.  Are you both members of the corp googlegroup or whatever you need to do to make CLA system happy?

Do you both work for Jetstack?

thockin commented:
Also please double-check the email address on all these commits.

agonzalezro commented:
Hi Tim, mine is signed as personal: agonzañezro@gmail.com

When it was only one commit (mine) the CLA was check was ok, when we added
the other commit from Matt is started failing.

Thanks for looking into this.

Regards,
Alex
On 29 Sep 2015 4:41 pm, "Tim Hockin" <notifications@github.com> wrote:

> Also please double-check the email address on all these commits.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/14328#issuecomment-144097595>
> .
>

thockin commented:
Maybe Saad can take a look and then you can squash - CLAbot seems to be happy with you (Sorry Matt :)

I took just a quick glance at the last 2 commits and they look OK, but I don't have full context.

@saad-ali I think it's do-or-die soon.

thockin commented:
@agonzalezro Travis is complaining about your godeps not being correct.  You have to make Travis happy.

agonzalezro commented:
@thockin there is a problem with the godeps checker: https://github.com/kubernetes/kubernetes/issues/14677

The issue is still open and it doesn't look like there is any workaround it yet.

thockin commented:
Sorry, I saw that after responding here.  watching that, too

On Tue, Sep 29, 2015 at 8:59 AM, Alexandre González <
notifications@github.com> wrote:

> @thockin <https://github.com/thockin> there is a problem with the godeps
> checker: #14677 <https://github.com/kubernetes/kubernetes/issues/14677>
>
> The issue is still open and it doesn't look like there is any workaround
> it yet.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/14328#issuecomment-144102774>
> .
>

fabioy commented:
I've manually confirmed the CLA, so changing the tags. Now waiting for green checks...

fabioy commented:
@k8s-bot please test

saad-ali commented:
Taking a look

saad-ali commented on docs/user-guide/volumes.md:
```diff
@@ -317,6 +319,17 @@ simultaneous readers allowed.
+[Flocker](https://clusterhq.com/flocker) is an open-source clustered container data volume manager. It provides management
+and orchestration of data volumes backed by a variety of storage backends.
+
+A `flocker` volume allows a Flocker dataset to be mounted into a pod. If the dataset does not already exist in Flocker, it is automatically provisioned, attached and mounted. Otherwise, with an existing dataset, it is reattached by Flocker to the node that the pod is scheduled. This means data can be "handed off" between pods as required.
```
Please update this (`...automatically provisioned...`)

saad-ali commented on pkg/volume/flocker/plugin.go:
```diff
@@ -0,0 +1,214 @@
+
+			select {
+			case <-timeoutChan:
+				return err
```
There is a possibility that `err` could be nil here right?

agonzalezro replied:
Not with the timeouts we have. It would already be calling line 191 so te `err` would be set to something. I can write my own error here if it helps, but I don't think that it's needed.

saad-ali replied:
I was thinking about the case "`err == nil`" and "`s.Primary != primaryUUID`"` and the timer tripped. I'm guessing that's unlikely?

agonzalezro replied:
That's pretty unlikely, I don't think we should even consider it.

saad-ali replied:
Doesn't hurt to be proactive. Let's wrap the err: instead of `return err` do `return fmt.Errorf("Timed out waiting for... %v", err)`

saad-ali commented on pkg/volume/flocker/plugin.go:
```diff
@@ -0,0 +1,214 @@
+
+		for {
+			if s, err = c.GetDatasetState(datasetID); err == nil && s.Primary == primaryUUID {
+				b.flocker.path = s.Path
```
nit: Since this code is identical to lines 207-209 below, it would be nice if it just broke out of the loop and executed that.

agonzalezro replied:
Good point

agonzalezro replied:
Actually... I think that there is a small error in this flow. I should `break` here as you mentioned. And continue on line 199. Does it sound good to you?

agonzalezro replied:
Mmmm not pretty sure that your solution is going to work. I am happy to hear any other advices.

saad-ali replied:
I'm sorry, I don't follow. Line 199 is `return err` for the time out. How/why would you want to go there?

I assume this (`primaryUUID` has been updated) is the terminating state, therefore you should break here which would break the loop and start executing at line 207 (`b.flocker.path = s.Path...`), which is what we want, right?

saad-ali commented:
Thanks for all your hard work on this @agonzalezro @mattbates 

LGTM mod a couple of nits

Once you've addressed them feel free to squash the changes and rebase (hopefully that'll make the CLA bot and tests happy).

mattbates commented:
Thanks @saad-ali for the feedback. We'll fix the nits where we can, re-test and then squash the changes and rebase as you suggest. EOD here so this will be updated early tomorrow now - and we're then over the line, hopefully!

thockin commented:
Lawyers say CLA is A-OK on this one.

Need to get CI happy still - rebase again?

saad-ali commented:
@k8s-bot ok to test

fabioy commented:
@k8s-bot please test

fabioy commented:
@saad-ali I think the CLA and test issues have been resolved. Could you LGTM/ok-to-merge it? Thanks.

saad-ali commented on pkg/volume/flocker/plugin.go:
```diff
@@ -0,0 +1,229 @@
+
+		select {
+		case <-timeoutChan:
+			return err
```
nit, but I'd still like to see this addressed:

There is a possibility that `err` will be nil, so please wrap the `err`: instead of `return err` do `return fmt.Errorf("Timed out waiting for... %v", err)`

I'm a fan of defensive programming.

saad-ali commented:
One last nit otherwise it looks good.

agonzalezro commented:
Hi @saad-ali I've updated the code adding your explicit `err`. Hoping that it looks ok now :) Thanks for the comments!

pmorie commented on pkg/api/types.go:
```diff
@@ -204,6 +204,9 @@ type VolumeSource struct {
 	// CephFS represents a Cephfs mount on the host that shares a pod's lifetime
 	CephFS *CephFSVolumeSource `json:"cephfs,omitempty"`
 
+	// Flocker represents a Flocker volume attached to a kubelet's host machine and exposed to the pod for its usage. This depends on the Flocker control service being running
```
Very minor nit, "and exposed to the pod for its usage" is redundant, since all volumes have that aspect :)

pmorie commented on pkg/api/types.go:
```diff
@@ -239,6 +242,8 @@ type PersistentVolumeSource struct {
 	CephFS *CephFSVolumeSource `json:"cephfs,omitempty"`
 	// FC represents a Fibre Channel resource that is attached to a kubelet's host machine and then exposed to the pod.
 	FC *FCVolumeSource `json:"fc,omitempty"`
+	// Flocker represents a Flocker volume attached to a kubelet's host machine and exposed to the pod for its usage. This depends on the Flocker control service being running
```
Same very minor nit here.

pmorie commented on pkg/api/v1/types.go:
```diff
@@ -250,6 +250,9 @@ type VolumeSource struct {
 	// CephFS represents a Ceph FS mount on the host that shares a pod's lifetime
 	CephFS *CephFSVolumeSource `json:"cephfs,omitempty"`
 
+	// Flocker represents a Flocker volume attached to a kubelet's host machine and exposed to the pod for its usage. This depends on the Flocker control service being running
```
Same nit

pmorie commented on pkg/api/v1/types.go:
```diff
@@ -589,6 +594,12 @@ type CephFSVolumeSource struct {
 	ReadOnly bool `json:"readOnly,omitempty"`
 }
 
+// FlockerVolumeSource represents a Flocker volume attached to a kubelet's host machine and exposed to the pod for its usage.
```
Same nit.  How about:

`FlockerVolumeSource represents a Flocker volume mounted by the Flocker agent`?

pmorie commented:
LGTM mod the very minor nits about comments.

agonzalezro commented:
Thanks for your comments @pmorie. Everything is updated now. Hope that it looks good.

pmorie commented:
LGTM, will leave to @saad-ali to tag

saad-ali commented:
LGTM. Thanks @agonzalezro and @mattbates 
Ok to merge on green

agonzalezro commented:
Hi, what's the problem with Jenkins? Everything looks ok in my end.

saad-ali commented:
```
15:15:27 Verifying ./hack/../hack/verify-generated-docs.sh
15:15:28 +++ [0930 15:15:28] Building go targets for linux/amd64:
15:15:28     cmd/gendocs
15:15:28     cmd/genman
15:15:28     cmd/genbashcomp
15:15:28     cmd/mungedocs
15:15:38 +++ [0930 15:15:38] Placing binaries
15:16:04 Some md files are missing ga-beacon analytics link:
15:16:04 /jenkins-master-data/jobs/kubernetes-pull-build-test-e2e-gce/workspace@3/_gopath/src/github.com/jstemmer/go-junit-report/README.md
15:16:04 /jenkins-master-data/jobs/kubernetes-pull-build-test-e2e-gce/workspace@3/_gopath/src/github.com/tools/godep/Readme.md
15:16:04 /jenkins-master-data/jobs/kubernetes-pull-build-test-e2e-gce/workspace@3/_gopath/src/github.com/tools/godep/Changelog.md
15:16:04 Generated docs are out of date. Please run hack/update-generated-docs.sh
15:16:04 !!! Error in ./hack/../hack/verify-generated-docs.sh:28
15:16:04   '"${KUBE_ROOT}/hack/after-build/verify-generated-docs.sh" "$@"' exited with status 1
15:16:04 Call stack:
15:16:04   1: ./hack/../hack/verify-generated-docs.sh:28 main(...)
15:16:04 Exiting with status 1
15:16:04 FAILED
15:16:04 Verifying ./hack/../hack/verify-generated-swagger-docs.sh
15:16:07 +++ [0930 15:16:07] Building go targets for linux/amd64:
15:16:07     cmd/genswaggertypedocs
15:16:08 +++ [0930 15:16:08] Placing binaries
15:16:10 Generating swagger type docs for unversioned
15:16:12 Generating swagger type docs for v1
15:16:13 Generating swagger type docs for experimental/v1alpha1
15:16:20 +++ [0930 15:16:20] Building go targets for linux/amd64:
15:16:20     cmd/kube-apiserver
15:17:30 +++ [0930 15:17:30] Placing binaries
15:17:31 
15:17:31 etcd appears to already be running on this machine (8723 etcd.test
15:17:31 8992 etcd.test
15:17:31 9020 etcd.test
15:17:31 9056 etcd.test) (or its a zombie and you need to kill its parent).
15:17:31 
15:17:31 
15:17:31 retry after you resolve this etcd error.
15:17:31 
15:17:31 +++ [0930 15:17:31] Clean up complete
15:17:31 !!! Error in /jenkins-master-data/jobs/kubernetes-pull-build-test-e2e-gce/workspace@3/hack/update-swagger-spec.sh:31
15:17:31   '"${KUBE_ROOT}/hack/after-build/update-swagger-spec.sh" "$@"' exited with status 1
15:17:31 Call stack:
15:17:31   1: /jenkins-master-data/jobs/kubernetes-pull-build-test-e2e-gce/workspace@3/hack/update-swagger-spec.sh:31 main(...)
15:17:31 Exiting with status 1
15:17:31 !!! Error in /jenkins-master-data/jobs/kubernetes-pull-build-test-e2e-gce/workspace@3/hack/update-generated-swagger-docs.sh:68
15:17:31   '"${KUBE_ROOT}/hack/update-swagger-spec.sh"' exited with status 1
15:17:31 Call stack:
15:17:31   1: /jenkins-master-data/jobs/kubernetes-pull-build-test-e2e-gce/workspace@3/hack/update-generated-swagger-docs.sh:68 main(...)
15:17:31 Exiting with status 1
15:17:31 !!! Error in /jenkins-master-data/jobs/kubernetes-pull-build-test-e2e-gce/workspace@3/hack/after-build/verify-generated-swagger-docs.sh:46
15:17:31   '"${KUBE_ROOT}/hack/update-generated-swagger-docs.sh"' exited with status 1
15:17:31 Call stack:
15:17:31   1: /jenkins-master-data/jobs/kubernetes-pull-build-test-e2e-gce/workspace@3/hack/after-build/verify-generated-swagger-docs.sh:46 main(...)
15:17:31 Exiting with status 1
15:17:31 !!! Error in ./hack/../hack/verify-generated-swagger-docs.sh:28
15:17:31   '"${KUBE_ROOT}/hack/after-build/verify-generated-swagger-docs.sh" "$@"' exited with status 1
15:17:31 Call stack:
15:17:31   1: ./hack/../hack/verify-generated-swagger-docs.sh:28 main(...)
15:17:31 Exiting with status 1
15:17:31 FAILED

```

agonzalezro commented:
@saad-ali I've repushed. But TBH I doubt that anything I have changed could cause the error that you pasted. Let's wait for the checks, but it should be ok! (I hope so :) Thanks!

saad-ali commented:
Make sure to re-run `hack/update-generated-docs.sh`

agonzalezro commented:
@saad-ali I've run that & `hack/update-all.sh` just to be sure and nothing have changed

agonzalezro commented:
Hello @saad-ali, just wondering if you need something else from me to add the `LGTM` label again.

saad-ali commented:
LGTM re-added.
We'll let the mergebot pick this up.

thockin commented:
Manually merging
